File: tests/fixtures/ts/mutation.ts
// ventouse TS fixtures — Mutation with type precision (P1). `readonly`/`Readonly<T>` params can't
// be mutated by the callee (compiler-enforced, like C++ `const`) -> never input mutation.

interface Box {
  x: number;
}

// TS-MUT-ARR — a mutable array param mutated -> dirty 20
function pushIt(arr: number[], x: number): void {
  arr.push(x);
}

// TS-MUT-OBJ — a mutable object param field written -> dirty 20
function setX(o: Box, v: number): void {
  o.x = v;
}

// TS-MUT-READONLY-ARR — `readonly number[]` cannot be mutated -> clean 0
function total(arr: readonly number[]): number {
  return arr.reduce((a, b) => a + b, 0);
}

// TS-MUT-READONLY-OBJ — `Readonly<Box>` param cannot be mutated -> clean 0
function getX(o: Readonly<Box>): number {
  return o.x;
}

// TS-MUT-LOCAL — local owned array, mutated locally -> clean 0
function build(n: number): number[] {
  const a: number[] = [];
  for (let i = 0; i < n; i++) {
    a.push(i);
  }
  return a;
}

File: tests/fixtures/rust/mutation.rs
// ventouse Rust fixtures — Mutation (P1). Signature-precise: only `&mut` can mutate the caller.

// RS-MUT-REF — write through a &mut param -> dirty 20
fn set_ref(p: &mut i32) { *p = 5; }

// RS-MUT-SHARED — a &T param: the compiler forbids mutation -> never input mutation -> clean 0
fn read_ref(p: &i32) -> i32 { *p }

// RS-MUT-BYVAL — `mut v` BY VALUE is owned (moved in); mutating it is local -> clean 0
fn consume(mut v: Vec<i32>) -> usize { v.push(1); v.len() }

// RS-MUT-LOCAL — a local owned vec, mutated locally -> clean 0
fn build(n: i32) -> Vec<i32> {
    let mut out = Vec::new();
    for i in 0..n { out.push(i); }
    out
}
